### PR TITLE
新增週訊息量統計功能

### PR DESCRIPTION
**Pull request by m1a2st**
## Why need this change? / Root cause: 
- want count message volume in this week
## Changes made:
- new /weekly-message-volume command
## Test Scope / Change impact:
- none
#137 

### Summary by OpenAI

**New Features:**
- Added a new function `addRequiredOption` to the `SlashCommandData` class for adding required options to slash commands.
- Introduced a new module "weekly-message-volume" with functionality to count weekly message volume in a specific channel using the `/weekly-message-volume` slash command.
- Implemented a `WeekRange` class to calculate the start and end dates of a weekly range.

**Refactor:**
- Replaced the `filter` function with `takeWhile` function in the `cherryPickMessages` function.

> 🎉 With every slash, a command takes flight, 🚀
> Counting messages, day and night. 🌞🌜
> A cherry-picked change, oh so slight, 🍒
> Makes our codebase shine bright! 💡🌟

### CODE REVIEW

File: discord-impl-jda/src/main/kotlin/tw/waterballsa/utopia/jda/extensions/GenericCommandInteractionEventExtension.kt [hunk 1]
```diff
@@ -75,3 +76,6 @@ fun SubcommandData.addRequiredOption(type: OptionType, name: String, description
     addOption(type, name, description, true)
 
 fun IReplyCallback.replyEphemerally(message: String) = reply(message).setEphemeral(true).queue()
+
+fun SlashCommandData.addRequiredOption(type: OptionType, name: String, description: String) =
+    addOption(type, name, description, true)
```
Comment (Wally5077) [inline discord-impl-jda/src/main/kotlin/tw/waterballsa/utopia/jda/extensions/GenericCommandInteractionEventExtension.kt:81]: **Well done**

File: weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import tw.waterballsa.utopia.jda.extensions.replyEphemerally
+import tw.waterballsa.utopia.weeklymessagevolume.WeeklyMessageVolumeListener.Companion.CHANNEL_NAME
+import tw.waterballsa.utopia.weeklymessagevolume.doamin.WeeklyRange
+
+@Component
+class WeeklyMessageVolumeListener : UtopiaListener() {
+
+    companion object {
+        private const val WEEKLY_MESSAGE_VOLUME = "weekly-message-volume"
+        const val CHANNEL_NAME = "channel-name"
```
Comment (Wally5077) [inline weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt:21]: 1. `private`
2. 有用到這個變數的擴充，放進 **WeeklyMessageVolumeListener**  裡

File: weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            Commands.slash(WEEKLY_MESSAGE_VOLUME, "Show the weekly message volume of the channel")
+                .addRequiredOption(OptionType.STRING, CHANNEL_NAME, "The channel to show the weekly message volume")
+        )
+    }
+
+    override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent) {
+        with(event) { countWeeklyMessage() }
+    }
+
+    private fun SlashCommandInteractionEvent.countWeeklyMessage() {
```
Comment (Wally5077) [inline weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt:35]: 邏輯放進 **onSlashCommandInteraction** 裡就好，這函數是多開的

File: weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        val messageVolume = textChannel.countMessageInWeeklyRange()
+
+        replyEphemerally("The weekly message volume of the channel  ${textChannel.name} is $messageVolume messages.")
+    }
+}
+
+private fun SlashCommandInteractionEvent.getChannelName() =
+    getOptionAsStringWithValidation(CHANNEL_NAME, "channel name is invalid.") { it.isNotBlank() }!!
```
Comment (Wally5077) [inline weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt:53]: 改成擴充屬性

File: weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+private fun SlashCommandInteractionEvent.getTextChannelsByNameIgnoreCase(): TextChannel? =
+    guild?.textChannels?.first { it.name.contains(getChannelName(), ignoreCase = true) }
```
Comment (Wally5077) [inline weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt:56]: **first** -> **find**

File: weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/doamin/WeeklyRange.kt [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+                .toTaiwanDateTime()
+    }
+
+    fun inWeeklyRange(time: OffsetDateTime): Boolean =
+        startDateTime.isBefore(time) && time.isBefore(endDateTime)
+}
+
+private fun LocalDate.toTaiwanDateTime(): OffsetDateTime =
+    atStartOfDay().atZone(ZoneId.of("Asia/Taipei"))
+        .toOffsetDateTime()
```
Comment (Wally5077) [inline weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/doamin/WeeklyRange.kt:30]: 用成一行

File: weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/doamin/WeeklyRange.kt [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    constructor() : this(getStartDateTime(), getEndDateTime())
+
+    companion object {
+        private fun getStartDateTime(): OffsetDateTime =
+            now().minusDays(7)
+                .toTaiwanDateTime()
+
+        private fun getEndDateTime(): OffsetDateTime =
+            now().plusDays(1)
+                .toTaiwanDateTime()
```
Comment (Wally5077) [inline weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/doamin/WeeklyRange.kt:21]: 一周不是七天前的概念，一周的起始是周一
假如今天周四，那麼時間回溯到 7 天前
就會變成**上**周四，這就不是當週了吧？

因此，就命名上會有語意上的錯誤

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

File: weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent) {
+        with(event) { countWeeklyMessage() }
+    }
+
+    private fun SlashCommandInteractionEvent.countWeeklyMessage() {
+        if (fullCommandName != WEEKLY_MESSAGE_VOLUME) {
+            return
+        }
```
Comment (frankvicky) [inline weekly-message-volume/src/main/kotlin/tw/waterballsa/utopia/weeklymessagevolume/WeeklyMessageVolumeListener.kt:38]: 合併

**Review by frankvicky — CHANGES_REQUESTED** [review]

**Review by Wally5077 — DISMISSED** [review]
**LGTM**

**Review by frankvicky — APPROVED** [review]

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Pull request merged**